File: rust/src/solution283.rs
use crate::common::Solution;

impl Solution {
    pub fn move_zeroes(nums: &mut Vec<i32>) {
        // slow indexes indicating non zero values, but not including the current index
        let mut slow = 0;
        let mut fast = 0;

        while fast < nums.len() {
            if nums[fast] == 0 {
                fast += 1;
                continue;
            }

            nums[slow] = nums[fast];
            slow += 1;
            fast += 1;
        }

        while slow < nums.len() {
            nums[slow] = 0;
            slow += 1;
        }
    }
}

File: java/Solution283.java
class Solution {
    public void moveZeroes(int[] nums) {
        // two pointer approach
        // p1 indicates the current non-zero areas not including p1
        // p2 is the scaning area, p2 will alawys stop at non-zero num

        int p1 = 0;
        for (int p2 = 0; p2 < nums.length; p2++) {
            if (nums[p2] == 0) continue;

            nums[p1++] = nums[p2];
        }

        while (p1 < nums.length) nums[p1++] = 0;
    }
}